src/main/java/com/bobocode/hoverla/bring/context/ClassBasedBeanDefinition.java

```diff
@@ -0,0 +1,265 @@
+     */
+    @Override
+    public void instantiate(BeanDefinition... dependencies) {
+        if (!isInstantiated()) {
+            log.debug("Creating new instance of bean '{}'", name);
+            instance = createInstance(dependencies);
+        }
+    }
+
+    /**
+     * See {@link BeanDefinition#getInstance()}.
+     *
+     * @return bean instance of current {@link BeanDefinition}.
+     * @throws NullPointerException when {@link ClassBasedBeanDefinition#instance} is null.
+     */
+    @Override
+    public Object getInstance() {
+        return Objects.requireNonNull(instance, "Instance of %s has not been created yet".formatted(name));
+    }
+
+    private String resolveName(Class<?> beanClass) {
+        Bean annotation = beanClass.getAnnotation(Bean.class);
+        String beanName = annotation.name();
+        if (isBlank(beanName)) {
+            return beanClass.getName();
+        }
+        return beanName;
+    }
+
+    private Map<String, Class<?>> resolveDependencies(Class<?> beanClass) {
+        Map<String, Class<?>> resolvedDependencies = new HashMap<>();
+        resolvedDependencies.putAll(resolveConstructorDependencies(beanClass));
+        resolvedDependencies.putAll(resolveFieldDependencies(beanClass));
+
+        return resolvedDependencies;
+    }
+
+    private Map<String, Class<?>> resolveConstructorDependencies(Class<?> beanClass) {
+        Constructor<?>[] beanConstructors = beanClass.getConstructors();
+
+        // we assume that at this point we should have only 1 @Inject constructor
+        // see BeanAnnotationClassValidator.java
+        Optional<Constructor<?>> injectionConstructor = Arrays.stream(beanConstructors)
+                .filter(ctor -> ctor.isAnnotationPresent(Inject.class))
+                .findFirst();
+        if (injectionConstructor.isPresent()) {
+            return resolveConstructorDependencies(injectionConstructor.get());
+        }
+
+        // we also assume that we should have only 1 plain constructor if @Inject constructor is missing
+        Constructor<?> plainConstructor = beanConstructors[0];
+        return resolveConstructorDependencies(plainConstructor);
+    }
+
+    private Map<String, Class<?>> resolveConstructorDependencies(Constructor<?> beanConstructor) {
+        this.constructor = beanConstructor;
+
+        Parameter[] constructorParameters = constructor.getParameters();
+        if (isEmpty(constructorParameters)) {
+            return emptyMap();
+        }
+        return Arrays.stream(constructorParameters)
+                .collect(toMap(param -> resolveMemberName(param, param.getType()), Parameter::getType));
+    }
+
+    private Map<String, Class<?>> resolveFieldDependencies(Class<?> beanClass) {
+        this.injectionFields = Arrays.stream(beanClass.getDeclaredFields())
+                .filter(field -> field.isAnnotationPresent(Inject.class))
+                .toList();
+
+        if (injectionFields.isEmpty()) {
```

Comment:
Won't stream return empty map for us?

Reply:
it will, but if we do this in a single chain,  we won't be able to initialize `injectionFields` variable, which is used later on